flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    @Test
+    public void testSuspend() {
+        MockContext ctx = new MockContext();
+        ctx.setHasEnoughResources(() -> false);
+
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
+        wfr.onEnter();
+        wfr.suspend(new Exception("test"));
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.FINISHED, JobStatus.SUSPENDED)));
+    } */
+
+    private class StateTestContext {
+        private final MockContext mockContext;
+        private WaitingForResources waitingForResources;
+
+        public StateTestContext(Consumer<MockContext> modifyContext) {
+            this.mockContext = new MockContext();
+            modifyContext.accept(mockContext);
+            this.waitingForResources =
+                    new WaitingForResources(mockContext, log, ResourceCounter.empty());
+            this.mockContext.setWaitingForResources(waitingForResources);
+            waitingForResources.onEnter();
+        }
+
+        public MockContext getMockContext() {
+            return mockContext;
+        }
+
+        public void assertStateTransitionTo(Class<? extends State> expected) {
+            Assert.assertThat(mockContext.getTransitionTo(), instanceOf(expected));
+        }
+
+        public void assertNoStateTransition() {
+            Assert.assertThat(mockContext.getTransitionTo(), nullValue());
+        }
+
+        public WaitingForResources getWaitingForResources() {
+            return waitingForResources;
+        }
+    }
+
+    private static class MockContext implements WaitingForResources.Context {
+
+        private static final Logger LOG = LoggerFactory.getLogger(MockContext.class);
+
+        private WaitingForResources waitingForResources;
+
+        private State transitionTo = null;
+        private Supplier<Boolean> hasEnoughResourcesSupplier = () -> false;
+        private Supplier<ExecutionGraph> createExecutionGraphWithAvailableResources =
+                () -> {
+                    try {
+                        return TestingExecutionGraphBuilder.newBuilder().build();
+                    } catch (Throwable e) {
+                        throw new RuntimeException("Error", e);
+                    }
+                };
+        private final List<Tuple3<State, Runnable, Duration>> actions = new ArrayList<>();
+
+        private void transitionToState(State target) {
+            if (transitionTo != null) {
+                Assert.fail("More than one state transition initiated");
+            }
+            transitionTo = target;
+            waitingForResources.onLeave(target);
+        }
+
+        private State getTransitionTo() {
+            return transitionTo;
+        }
+
+        @Override
+        public void goToFinished(ArchivedExecutionGraph archivedExecutionGraph) {
+            transitionToState(new Finished(null, archivedExecutionGraph, LOG));
```

Comment:
I'd say that this goes too far. These are scheduler implementation details (e.g., how the `Finished` state is represented), or what happens within `goToFinished` in it's entirety.

This path will likely lead us toward just another state machine.

Reply:
I was thinking more of something like this:

```
class MockContext ... implements AutoCloseable

void setExpectFinished(Consumer<EG> asserter) {
    this.noFinishTrap=() -> throw new AssertionError("no transition to finished");
    this.setGoToExecutingConsumer(asserter);
}

void goToExecuting(EG eg) {
   this.noFinishTrap = ()->{};
   // assert arguments
   this.goToExecutinConsumer.accept(eg);
}

void close() {
   // noop if expected transition occurred
   this.noFinishTrap.run();
}
```

Reply:
> I'd say that this goes too far. These are scheduler implementation details (e.g., how the Finished state is represented), or what happens within goToFinished in it's entirety.

I would have really liked to avoid this, but I added it because we need a `Finished` instance to pass to the `onLeave(State)` method.

Reply:
Also thanks for the code mockup. What I like about it is that the tested class is instantiated in the test itself. My approach hides everything, and reduces the test only to the required input (making the tests shorter, but also more abstract).

I'll try to implement a few tests with your approach to see how well it works

Reply:
Oh, I didn't even notice that onLeave gets the _new state_ ... ideally we get rid of that. Maybe we can tie to disposal of the operator coordinator to the EG termination future instead.

Reply:
`onLeave(State)` is indeed not optimal. This somewhat couples those implementations which rely on a specific State implementation in the `onLeave` method. Maybe we can test this method in isolation and with more of an integration test.

The problem `onLeave(State)` should solve is that we need to close some services when leaving a certain state. One way to solve it could be to pass this service to the context method and let the implementor of this context decide how to handle the service. The downside is that now every state transition would get the full set of services.

Alternatively, the state might need to remember into which state it transitioned and depending on this close or leave the service when `onLeave()` is called (w/o the new state). But this seems brittle.

I think that Chesnay's proposal should work.